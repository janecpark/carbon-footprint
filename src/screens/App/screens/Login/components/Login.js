import React, { useState, useContext } from 'react';
import UserApi from '../../../UserApi';
import TokenContext from '../../../TokenContext';
import { useHistory } from 'react-router-dom';
import Alert from '../../../../../shared/components/Alert';

/** Log in form that logs in users and sets the token in localStorage */

const Login = () => {
  const history = useHistory();
  const { setToken } = useContext(TokenContext);
  const [userData, setUserData] = useState({
    username: '',
    password: '',
    errors: [],
  });

  const handleChange = (e) => {
    const { name, value } = e.target;
    setUserData((el) => ({
      ...el,
      [name]: value,
    }));
  };
  async function handleSubmit(e) {
    e.preventDefault();
    let data = {
      username: userData.username,
      password: userData.password,
    };
    try {
      const res = await UserApi.login(data);
      setToken(res.token);
      history.push('/');
    } catch (err) {
      setUserData((f) => ({ ...f, errors: [...err] }));
    }
  }

  const loginForm = (
    <form className="container login-form" onSubmit={handleSubmit}>
      <div className="form-group">
        <label htmlFor="username">Username</label>
        <input
          type="text"
          className="form-control"
          name="username"
          value={userData.username}
          onChange={handleChange}
          required
        />
      </div>
      <div className="form-group">
        <label htmlFor="password">Password</label>
        <input
          type="password"
          className="form-control"
          name="password"
          value={userData.password}
          onChange={handleChange}
          required
        />
      </div>
      <button className="btn btn-primary btn-block">Log in</button>
    </form>
  );

  return (
    <div className="Login">
      <h2 className='Login-header'>Login</h2>
      {userData.errors.length ? (
        <Alert type="danger" messages={userData.errors} />
      ) : null}
      {loginForm}
    </div>
  );
};

export default Login;
