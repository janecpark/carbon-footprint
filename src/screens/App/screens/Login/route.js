import { routeConstants } from '../../../../shared/constants';
import Login from "./";

export default {
    path: routeConstants.LOGIN.route,
    component: Login
};