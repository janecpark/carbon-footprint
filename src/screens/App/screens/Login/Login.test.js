import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import Login from './components/Login';
import {
  UserProvider,
  FormProvider,
  ResultProvider,
  ActionProvider,
  TokenProvider,
} from '../testUtils';

it('renders without crashing', function () {
  render(
    <MemoryRouter>
      <UserProvider>
        <FormProvider>
          <ResultProvider>
            <ActionProvider>
              <TokenProvider>
                <Login />
              </TokenProvider>
            </ActionProvider>
          </ResultProvider>
        </FormProvider>
      </UserProvider>
    </MemoryRouter>
  );
});

it('matches snapshot', function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <FormProvider>
          <ResultProvider>
            <ActionProvider>
              <TokenProvider>
                <Login />
              </TokenProvider>
            </ActionProvider>
          </ResultProvider>
        </FormProvider>
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

test('renders text on page', () => {
  const { getByText } = render(
    <MemoryRouter>
      <UserProvider>
        <FormProvider>
          <ResultProvider>
            <ActionProvider>
              <TokenProvider>
                <Login />
              </TokenProvider>
            </ActionProvider>
          </ResultProvider>
        </FormProvider>
      </UserProvider>
    </MemoryRouter>
  );
  const element = getByText(/Username/);
  expect(element).toBeInTheDocument();
});
