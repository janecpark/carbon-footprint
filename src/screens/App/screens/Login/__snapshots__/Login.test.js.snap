// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Login"
  >
    <form
      class="container"
    >
      <div
        class="form-group"
      >
        <label
          for="username"
        >
          Username
        </label>
        <input
          class="form-control"
          name="username"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="password"
        >
          Password
        </label>
        <input
          class="form-control"
          name="password"
          required=""
          type="password"
          value=""
        />
      </div>
      <button
        class="btn btn-primary btn-block"
      >
        Log in
      </button>
    </form>
  </div>
</DocumentFragment>
`;
